### ISSUE TITLE
False positive for `trailing_closure` rule with anonymous closure call

### ISSUE THREAD

**Issue by proxi** [realm/SwiftLint#2159]
### New Issue Checklist

- [✅] Updated SwiftLint to the latest version
- [✅] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

It is sometimes handy to use anonymous closure as code grouping, e.g.:

```swift
menu.addItem({
    let menuItem = NSMenuItem(title: "Blah!")
    return menuItem
}())
```

When linted with `trailing_closure` enable, this incorrectly triggers a warning.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
Bagombo:closure proxi$ swiftlint 
Loading configuration from '.swiftlint.yml'
Linting Swift files in current working directory
Linting 'Closure.swift' (1/1)
Closure.swift:2:5: warning: Trailing Closure Violation: Trailing closure syntax should be used whenever possible. (trailing_closure)
Done linting! Found 1 violation, 0 serious in 1 file.
Bagombo:closure proxi$ 
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.25.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)? CocoaPods
* Paste your configuration file:

```yml
opt_in_rules:
    - trailing_closure
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents. No
* Which Xcode version are you using (check `xcode-select -p`)? 9.3
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
// This triggers a violation:
func foo(_ bar: Int) {
    foo({ return 42 }())
}
```

**Comment by Pashcal** [comment]
Also looks like a similar bug:
"trailing_closure" is triggering on `Dictionary(grouping: list, by: { $0.group }).sorted { $0.key < $1.key }`, but not with `Dictionary(grouping: list) { $0.group }.sorted { $0.key < $1.key }`.
For me, first options is more readable.

**Comment by marcelofabri** [comment]
@Pashcal for future reference: what you're describing is not bug: the rule checks all closures that can be closure. 

If you believe the first one is more readable, please open an issue to track adding a configuration to treat it as a special case. Please try to specify the enhancement as a generic case. When should the rule ignore the fact that a closure can be a trailing closure?
